File: seed/app/src/main/java/com/WATER_ME_WITH_COMPANY_GOODNESS/android/WATER_ME_WITH_APP_NAME/fragment/PresentableFragment.java
package com.{company_name}.android.{app_package_name_prefix}.fragment;

import android.os.Bundle;

import com.{company_name}.android.{app_package_name_prefix}.module.AppServicesComponent;
import com.{company_name}.android.{app_package_name_prefix}.mvp.presenter.Presenter;

/**
 * Base class for all fragments which have a corresponding {@link Presenter} object
 *
 * @param <T> The type of presenter the fragment uses
 */
public abstract class PresentableFragment<T extends Presenter> extends BaseFragment {

    private T presenter;

    protected abstract T createPresenter(AppServicesComponent servicesComponent);

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        presenter = createPresenter(app.getAppServicesComponent());
        if (presenter == null) {
            throw new IllegalStateException("presenter == null");
        }
        presenter.onCreate(savedInstanceState);
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        presenter.onSaveInstanceState(outState);
    }

    @Override
    public void onStart() {
        super.onStart();
        presenter.onStart();
    }

    @Override
    public void onResume() {
        super.onResume();
        presenter.onResume();
    }

    @Override
    public void onPause() {
        presenter.onPause();
        super.onPause();
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.onStop();
    }

    @Override
    public void onDestroy() {
        presenter.onDestroy();
        presenter = null;
        super.onDestroy();
    }

    public final T getPresenter() {
        return presenter;
    }
}

File: garden/seeds/kotlin/app/src/main/kotlin/com/WATER_ME_WITH_COMPANY/android/WATER_ME_WITH_APP_NAME/fragment/PresentableFragment.kt
package com.{{company_name}}.android.{{app_package_name_prefix}}.fragment

import android.os.Bundle
import com.{{company_name}}.android.{{app_package_name_prefix}}.module.AppServicesComponent
import com.{{company_name}}.android.{{app_package_name_prefix}}.mvp.presenter.Presenter
import com.{{company_name}}.android.{{app_package_name_prefix}}.mvp.view.MvpView

/**
 * Base class for all fragments which have a corresponding [Presenter] object

 * @param T The type of presenter the fragment uses
 */
public abstract class PresentableFragment<V : MvpView, T : Presenter<V>> : BaseFragment() {

    lateinit var presenter : T

    protected abstract fun createPresenter(servicesComponent: AppServicesComponent, savedInstanceState: Bundle?): T

    override public fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        this.presenter = createPresenter(getApp().getAppServicesComponent(), savedInstanceState)
        presenter.onCreate(savedInstanceState)
    }

    override public fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        presenter.onSaveInstanceState(outState)
    }

    override public fun onStart() {
        super.onStart()
        presenter.onStart()
    }

    override public fun onResume() {
        super.onResume()
        presenter.onResume()
    }

    override public fun onPause() {
        presenter.onPause()
        super.onPause()
    }

    override public fun onStop() {
        super.onStop()
        presenter.onStop()
    }

    override public fun onDestroy() {
        presenter.onDestroy()
        super.onDestroy()
    }
}